RFC Merging Process (#4124)
Scille/parsec-cloud: docs/development/merging-process.md

for general tips and best practices on these steps.

## Opening a pull request

You open a pull request to ask collaborators for feedback on your changes:

- when you have little or no code but want to share some general ideas,
- when you’re stuck and need help or advice,
- when you’re ready for someone to review your work

If you want early feedback or advice before you complete your changes, **please
mark your pull request as a draft**.

### What makes a good pull request

When you create a pull request, please follow the following guidelines:

1. **Smaller is better**. You should include the minimal set of changes required
   to address the subject of the pull request.
   - If the issue you are working on requires changing several software
     components at once, take some time after the "job is done" to split those
     changes into several small, coherent, pull request.
   - This reduces the time and effort required to review and merge the pull
     request
2. Include a **meaningful title**, describing what is being solved at a high
   level. Pull request are scanned and triaged by people, a good title will save

>    some time otherwise spent reading detailed descriptions and understading what

Reviewer: `understanding`

Author: ```suggestion
some time otherwise spent reading detailed descriptions and understanding what
```

   is it about.
3. For non-trivial changes, **write a useful description** to explain your
   changes to the reviewer(s). See [How to write a useful description](#how-to-write-a-useful-description).
4. If the pull request addresses an issue, please **link the issue** in the
   description.
5. If some other pull requests should be merged before, **list dependencies** in
   the description.
6. Ensure the [commit history is sanitized](#how-to-keep-a-sanitized-commit-history).
7. It should be affected to the related categories listed in [Pull request categories](#pull-request-categories).
8. Complete [the Pull-request checklist](#the-pull-request-checklist).

### How to write a useful description

A useful description is one that give all the hints necessary for the
reviewer(s) to understand what is been changed and to reason about your
approach.

Here are some tips:

> - Give some **context**: Why this pull request exist ? Why it was created ?

Reviewer: `exists`

Author: ```suggestion
 Give some **context**: Why this pull request exists ? Why it was created ?
```

A pull request might affect different parts of the codebase in different ways.

The following table lists the categories that can be manually or automatically
assigned to a pull request.

> Some categories could be automaticaly detected: for example if the pull
> request modifies `.rs` files we can assign the `Rust` category to it.

| Category     | Meaning                                        |
| ------------ | ---------------------------------------------- |
| Bug          | The PR fixes a bug                             |
| Security     | The PR is related to a security bug            |
| Feature      | The PR adds a new feature                      |
| Typo         | The PR fixes a typo                            |
| Dependencies | The PR updates one or more dependencies        |
| Release      | The PR is a release                            |
| RFC          | The PR is related to the RFCs documentation    |
| Docs         | The PR is related to the user documentation    |
| Gui          | The PR is related to the GUI                   |
| Test         | The PR is related to the test codebase         |
| **PER CODE** |                                                |

> | CI           | The PR affect the CI                           |

Author: ```suggestion
| CI           | The PR affects the CI                          |
```

- [ ] Link any related issue in the description
   <!-- You can add `closes #<IssueID>` to close the issue when the PR is merged -->
- [ ] Link any dependent pull request in the description

If this PR adds a new feature or fixes a bug:

- [ ] I have added or updated relevant tests

If this PR modifies installation or updates packages:

- [ ] I have checked that installation works on all platforms (Windows, Linux, Mac ...)
    <!-- If you don't have a specific platform, seek help of someone that have it. -->

- [ ] I have updated related documentations (installation commands, fix to specific problems ...)

If this PR is related to GUI:

- [ ] I have checked that GUIs affected by my changes are working properly.

If this PR affects the user:

<!-- Changes that have an impact for the user:
     - UI has been changed
     - A message displayed to the user has been changed
     - A new feature has been added
     - A use case has been modified
     - ... -->

- [ ] I have created a news fragment
   <!-- The news fragments are used to generate the changelog -->
- [ ] I have updated the user documentation
```

## Reviewing a pull request

> Before a pull request can be merged, it need to be reviewed to meet our quality

Reviewer: `needs`

- Ensure that [commit history is sanitized](#how-to-keep-a-sanitized-commit-history)

#### The code owner

A *code owner* is a person or a team assigned to be responsible for a specific
component or code parts, usually because it has a deep knownledge about it.

> For example the Python dev are the owner of the python code.

### Specific pull request reviews

This section describes extra steps or things to pay attention to for specific
types of pull requests.

#### Pull request: fix typos

We want the codebase to not be cluttered with typos (syntax, grammar and others).
Pull requests that solves those issues are welcome!

When reviewing this type of pull request, ensure that:

- It does not introduce new typos
- The original meaning is not changed

Also, take the opportunity to consider rephrasing to better convey the initial
meaning.

> We don't want to introduce new typos when correcting some that are already present. (ouroboros)

#### Pull request: update dependencies

Likely opened by a bot (like dependabot) these are pull requests that update
one or more required dependencies and are small (few files are changed, most
likely one line per file).

> In addition to the standard process, we would like to inspect the change

Reviewer: `changes`

Author: ```suggestion
In addition to the standard process, we would like to inspect the changes
```

Here the burden remains on the reviewer (considering the update was created by a
bot). Take the time to read dependency changelog, blog posts, or similar. Keep
in mind that this step could prevent (to a certain limit) a potential
supply-chain attack where a fake update is released with malicious code.

## Merging a pull request

After the pull request has been reviewed and accepted, the author is allowed to
merge it into the main branch.

The merge should be done in a specific way:

- The commits should appear on top of the git history (see [Why do we want a linear git history](#why-do-we-want-a-linear-git-history)).
- It should be tested one last time if changes are compatible with the main
  branch.

    This is done by the [merge queue](#what-is-a-merge-queue) that will rebase the
    commits onto the main branch.

### Why do we want a linear git history

Why want a linear git history ? because it allows to:

- Answer the question: *Did **feature A** get introduced before of after **bugfix B** ?*
- Simpler git history graph (everything is in-line).
- `git bisect` work best with a linear history.

> - Remove unecessary commit `Merge branch X into Y` that clutter the history.

Author: ```suggestion
- Remove unnecessary commit `Merge branch X into Y` that clutters the history.
```

How to achieve a linear git history?

> Git provide multiple methods for that

Author: ```suggestion
Git provides multiple methods for that
```

- `git fast forward` (`git merge --ff-only`):

  When commits are already ahead of the base branch, it will only update the
  branch pointer.

- `git squash` (`git merge --squash`):

>   Apply the change onto the destination branch but wont make a commit, this

Author: ```suggestion
  Apply the changes onto the destination branch but wont make a commit, this
```

  allows to create a single commit on top of the current branch.

- `git rebase`

  Will replay commits that are not already on the `upstream` branch on top of
  the `upstream` branch's commit.

### What is a merge queue

A merge queue is a system that will queue pull requests to be merge in a queue :p

A merge queue typically works by creating a temporary branch that include:

- The latest change from the base branch.
- The change introduced by the PRs ahead of the queue.

> - The change added by your PR.

Reviewer: `changes`

Author: ```suggestion
- The changes added by your PR.
```

This ensures that we merge pull requests that work with the latest state of the

> base branch and the previous to be merge PR.

Author: The previous PRs that are already in the queue
